### PR TITLE
feat: status expression in components

### PR DESCRIPTION
**Pull request by adityathebe**
resolves: https://github.com/flanksource/canary-checker/issues/1620

### CODE REVIEW

File: pkg/system_api.go [hunk 1]
```diff
@@ -162,7 +162,9 @@ type Component struct {
 	CostTotal7d     float64                     `json:"cost_total_7d,omitempty" gorm:"column:cost_total_7d"`
 	CostTotal30d    float64                     `json:"cost_total_30d,omitempty" gorm:"column:cost_total_30d"`
 	LogSelectors    dutyTypes.LogSelectors      `json:"logs,omitempty" gorm:"column:log_selectors"`
-	ParentLookup    *v1.ParentLookup            `json:"parentLookup,omitempty" gorm:"-"`
+
+	ParentLookup *v1.ParentLookup `json:"parentLookup,omitempty" gorm:"-"`
+	StatusExpr   string           `json:"statusExpr,omitempty" gorm:"-"`
```
Comment (moshloop) [inline pkg/system_api.go:167]: Doesn't this need to be  saved on the db, so it can be run again on query ?

Comment (adityathebe) [reply pkg/system_api.go:167]: That is persisted on the topology spec.

File: fixtures/topology/component-status-expr.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+apiVersion: canaries.flanksource.com/v1
+kind: Topology
+metadata:
+  name: status-expr
+spec:
+  icon: flux
+  type: Topology
+  schedule: "@every 5m"
+  statusExpr: |
+    summary.healthy > 0 && summary.unhealthy == 0 ? "good" : "bad"
```
Comment (moshloop) [inline fixtures/topology/component-status-expr.yml:11]:
```suggestion
    summary.healthy > 0 && summary.unhealthy == 0 && summary.warning == 0 ? "healthy" : 
    summary.healthy == 0 && summary.unhealthy == 0 && summary.warning == 0 ? "unknown" :   
    summary.healthy > 0 && summary.unhealthy > 0 || summary.warning  > 0 ? "warning" :
    "unhealthy"
```

File: pkg/topology/run.go [hunk 3]
```diff
@@ -652,11 +654,24 @@ func (tj *TopologyJob) run(ctx *ComponentContext, job job.JobRuntime) (pkg.Compo
...
 		ctx.Infof(results.Debug(ctx.Logger.IsLevelEnabled(5), ""))
 	}
+
 	for _, c := range results.Walk() {
 		if c.Namespace == "" {
 			c.Namespace = ctx.Topology.GetNamespace()
 		}
 		c.Schedule = ctx.Topology.Spec.Schedule
+
+		if c.StatusExpr != "" {
```
Comment (moshloop) [inline pkg/topology/run.go:664]: This also need to happen during topology query

File: pkg/system_api.go (hunk 1 above)
Comment (adityathebe) [reply pkg/system_api.go:167]: https://github.com/flanksource/duty/pull/876

File: Makefile [hunk 4]
```diff
@@ -240,7 +240,7 @@ bin: .bin .bin/wait4x .bin/karina
 
 # Generate all the resources and formats your code, i.e: CRDs, controller-gen, static
 .PHONY: resources
-resources: fmt static manifests
```
Comment (moshloop) [inline Makefile:243]: Why are we removing this ?

Comment (adityathebe) [reply Makefile:243]: `static` calls `manifests`

**Pull request merged**